File: week2/B_9012.py
T = int(input())

for tc in range(T):
    ps = input()
    stack = []
    
    for n in ps:
        if n == '(':
            stack.append(n)
        elif n == ')':
            if len(stack) != 0 and stack[-1] == '(':
                stack.pop()
            else:
                stack.append(n)
                break
    
    if len(stack) == 0:
        print('YES')
    else:
        print('NO')

File: java/src/main/java/B_9012.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class B_9012 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int t = Integer.parseInt(br.readLine());

        for(int i = 0; i < t; i++) {
            String str = br.readLine();
            Stack<Character> stack = new Stack<Character>();

            for(int j = 0; j < str.length(); j++) {
                if(str.charAt(j) == '(') {
                    stack.push(str.charAt(j));
                } else {
                    if(stack.empty()) {
                        stack.push(str.charAt(j));
                        break;
                    } else {
                        stack.pop();
                    }
                }
            }

            if(stack.empty()) {
                sb.append("YES");
            } else {
                sb.append("NO");
            }
            sb.append('\n');
        }
        System.out.println(sb);
    }
}
